### PR TITLE
Suppress ``useless-super-delegation`` if return type changed (#5822)

### LINKED ISSUE #5822: useless-super-delegation when an override changes the return type annotation
### Current problem

Similar to #1923, but for the return type annotation.
It is sometimes useful to override a method and change the return type annotation - whether to help a static type checker out or to help the IDE make better autocomplete suggestions.

Here's an example - it's a bit contrived, I actually ran into this when overriding to add typing to a method from a third-party library (ANTLR).

```python
"""Quick test to demonstrate useless-super-declaration raised for overrides
that change return type annotations."""
import random
from dataclasses import dataclass
from typing import Any

# pylint: disable=too-few-public-methods
class _AGenerator:
    """Abstract base class for generators."""
    _valid_values: list[Any]

    def get_random(self) -> Any:
        """Generates a random value from the valid values that this generator
        can return."""
        return random.choice(self._valid_values)

@dataclass
class StrGenerator(_AGenerator):
    """Generator that accepts and returns string values."""
    _valid_values: list[str]

    # override to add typing
    def get_random(self) -> str:
        return super().get_random()

s = StrGenerator(['foo', 'bar', 'qux'])
print(s.get_random())
print(s.get_random())
print(s.get_random())
```

Running pylint with default config on it:
```
$ python -m pylint test.py 
************* Module test
test.py:24:4: W0235: Useless super delegation in method 'get_random' (useless-super-delegation)

------------------------------------------------------------------
Your code has been rated at 9.33/10 (previous run: 9.33/10, +0.00)
```

### Desired solution

pylint should not emit a warning in this case as the override has legitimate uses.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by timmartin**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

In the `useless-super-delegation` checker, check whether the return type annotation has been specialized in the derived class. If so, then the override is not useless and we should not issue the warning.

Closes #5822

### CODE REVIEW

File: tests/functional/u/useless/useless_super_delegation_py3.py
Comment (DanielNoord) [inline tests/functional/u/useless/useless_super_delegation_py3.py:51]: This annotation is incorrect, is that intentional?

Comment (timmartin) [reply tests/functional/u/useless/useless_super_delegation_py3.py]: Well spotted, that wasn't intentional.

While I was fixing that, I realized there's another case we need to think about:

    import typing
    from typing import Any

    class Base:
        def foo(self) -> Any:
            pass

    class Derived:
        def foo(self) -> typing.Any:
            return super().foo()

This will be harder to fix, and it looks like the existing rules around type hints on parameters don't cope with this case either, so I suggest this can be handled as a later fix to both parameter type hints and return type hint.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM.

**Comment by DanielNoord** [comment]
@cdce8p I'd like to get your input here. I'm not sure if this change is reasonable. We're not a type checker, but should we support this? You're probably the best person to judge this.

**Comment by tushar-deepsource** [comment]
I'd say this is a good change. It's a change that will lead to less false positives, and it's something that we can infer, so why not take advantage of it?

**Comment by Pierre-Sassoulas** [comment]
@cdce8p could you check this, if you have some time ? It's blocking the release of 2.13.6 :)

File: pylint/checkers/classes/class_checker.py
Comment (cdce8p) [inline pylint/checkers/classes/class_checker.py:1269]:
```py
            if (
                function.returns is not None
                and meth_node.returns is not None
                and meth_node.returns.as_string() != function.returns.as_string()
            ):
```

**Review by cdce8p — COMMENTED** [review]
I'm not fully convinced it is all that useful, but I guess we can add it. Especially since we do something similar for the parameter annotations already.

Just two suggestions:
* Do `not` return if one function doesn't have a return type (similar to parameter annotations).
* Move it below the `if called_annotations and overridden_annotations` block.

**Comment by timmartin** [comment]
I'm not sure what you mean by "Do not return if one function doesn't have a return type (similar to parameter annotations)." Currently my reasoning is that if the derived method adds a type missing from the base method we should not warn `useless-super-delegation`, but if the base method has the annotation and the derived method doesn't then we should warn. Is that what you're meaning?

**Comment by cdce8p** [comment]
> I'm not sure what you mean by "Do not return if one function doesn't have a return type (similar to parameter annotations)." Currently my reasoning is that if the derived method adds a type missing from the base method we should not warn `useless-super-delegation`, but if the base method has the annotation and the derived method doesn't then we should warn. Is that what you're meaning?

My idea was to be careful not to block too many errors. I.e. my suggestion would be to only ignore `useless-super-delegation` if **both** the parent and child method have function annotations and they differ.

File: pylint/checkers/classes/class_checker.py
Comment (cdce8p) [inline pylint/checkers/classes/class_checker.py:1305]:
```suggestion
                if function.returns is not None and (
                    meth_node.returns is None
                    or meth_node.returns.as_string() != function.returns.as_string()
                ):
```
To fix the tests, you would need to use the old condition.

For my https://github.com/PyCQA/pylint/pull/6141#issuecomment-1099627809, I suggested something a little bit different.
```suggestion
            if (
                function.returns is not None
                and meth_node.returns is not None
                and meth_node.returns.as_string() != function.returns.as_string()
            ):
```

**Review by cdce8p — COMMENTED** [review]
Please don't force push changes new here. That will make it difficult / impossible to see and compare old revisions.

**Comment by timmartin** [comment]
My thinking was that the most likely reason I could think to override a method just to change the return type annotation is that I was extending some library code that doesn't have type declarations and I wanted my library to have type declarations, in which case issuing a diagnostic isn't helpful. I'm not sure I care enough about this to prolong the discussion, though.

**Comment by cdce8p** [comment]
> My thinking was that the most likely reason I could think to override a method just to change the return type annotation is that I was extending some library code that doesn't have type declarations and I wanted my library to have type declarations, in which case issuing a diagnostic isn't helpful. I'm not sure I care enough about this to prolong the discussion, though.

Usually, I think it would be better if the original library added type hints itself. However, I do agree that it's not worth the discussion. If you think it would be helpful, I guess we could do that, too. Will leave that up to you.

**Comment by timmartin** [comment]
No, I'm done with this change. Merge it or close the PR as you see fit.

**Review by cdce8p — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
Thank you for the fix @timmartin !

**Pull request merged**
